### ISSUE TITLE
Fix defects in unique_any

### ISSUE THREAD

**Issue by kkaefer** [mapbox/mapbox-gl-native#11631]
Our implementation of `unique_any` has a few issues:

* We're [explicitly calling the destructor when moving a stack-allocate object](https://github.com/mapbox/mapbox-gl-native/blob/05d24448bdea29e302f788f050829459adba8e83/include/mbgl/util/unique_any.hpp#L159), which means that it's going to be called twice on the same memory: once explicitly, and once implicitly as part of the stack unwinding
* In that same function, we're always treating `src.stack` as valid memory. When the original `unique_any` was created without an initial value, this is undefined memory. This means, we're calling a destructor on undefined memory as we exit this function.
* The various `any_cast` functions use `static_assert` to ensure that `ValueType` is copyable. This means that despite the claim at the top of this file, you can't actually use move-only types with `unique_any`. Instead, we should use `enable_if` to explicitly enable these overloads iff `ValueType` is constructible

There is considerable complexity in this code to switch between stack and heap allocation based on the size of `ValueType`, but we're only ever using the stack allocated path in our codebase.

Instead, I propose to replace this with a type-erased object with a virtual destructor and `dynamic_cast` when accessing the "any" object in the bindings.

/cc @jfirebaugh @asheemmamoowala

**Comment by jfirebaugh** [comment]
> We're explicitly calling the destructor when moving a stack-allocate object

Although it does more work than necessary, the code is correct as written. It calls the move constructor twice: once for the stack allocated object, and once for the placement new. It therefore needs to call the destructor twice as well: once for the stack allocated object, and once for "placement delete" of the moved-from value.

It would be better written as:

```c++
new (&dest.stack) ValueType(std::move(reinterpret_cast<ValueType&>(src.stack)));
destroy(src);
```

> * In that same function, we're always treating `src.stack` as valid memory.

I don't believe this is an issue; calls to `vtable->move` are guarded with `if (vtable)`, and uninitialized values have a null `vtable`.

> * The various `any_cast` functions use `static_assert` to ensure that `ValueType` is copyable.

Not all of them do. The `static_assert` is present in the `const unique_any&` and `unique_any&` overloads of `any_cast`, but not the `unique_any*` or `unique_any&&` overloads. I believe there was a reason we opted to have the former require copyability rather than returning a reference, but I don't remember what it was.

**Comment by asheemmamoowala** [comment]
> > The various any_cast functions use static_assert to ensure that ValueType is copyable.
> 
> Not all of them do. The static_assert is present in the const unique_any& and unique_any& overloads of any_cast, but not the unique_any* or unique_any&& overloads. I believe there was a reason we opted to have the former require copyability rather than returning a reference, but I don't remember what it was.

The `any_cast` with copied return value is currenlty only used in the drawin Source peer: https://github.com/mapbox/mapbox-gl-native/blob/d176e9bb46c4ab539e6cc5f6c7d17bc68d95e205/platform/darwin/src/MGLStyle.mm#L180

I think this is a mistake, there is an easy fix for this, which is already used in the android bindings: 
https://github.com/mapbox/mapbox-gl-native/blob/d176e9bb46c4ab539e6cc5f6c7d17bc68d95e205/platform/android/src/style/sources/source.cpp#L54

**Comment by asheemmamoowala** [comment]
Fixed in #12166. 
Remaining reported issue was nixed in [#12215(comment)](https://github.com/mapbox/mapbox-gl-native/pull/12215#pullrequestreview-131830212)
